### PR TITLE
Refactor tmx entry

### PR DESCRIPTION
**Pull request by t-cordonnier**
Introduce interfaces to TMXEntry

These interfaces have multiple goals.

In short term, they enable to merge functions actually duplicated, as you can see in my second patch (in Searcher class)

In middle term (I did only one example in patch 3) they enable to create an immutable view to (maybe) mutable entries: now users of ExternalTMX cannot modify its contents. We could do the same for ProjectTMX: store mutable entries in the maps, but return ITMXEntry in getDefaultTranslation/getMultipleTranslation
As @amake mentionned in the discussion list, PrepareTMXEntry actually looks like mutable variant of TMXEntry. That was initial meaning, but if you look where they are used, you will see that since ExternalTMX continued to use PrepareTMXEntry, it has been used as something similar to external TMX entry.
One other difference between these two classes is that PrepareTMXEntry can contain any kind of properties (as it can be read from any kind of TMX) while TMXEntry stores only properties a project_save.tmx should contain (actually no property but already stores externalLinked): this enforces the idea that TMXEntry means, in reality, internal TMX entry.
Then we can ask ourselves: why does ExternalTMX store mutable entries while internal (project) TMX contains immutable entries? Shouldn't the contrary be more logical?

In long term, these two interfaces could ensure stability of the code even if you change the definition of TMXEntry/PrepareTMXEntry (as @miurahr wanted to do), even if you finally remove one of them or rewrite it completely: all code using the interfaces would not need to be changed. 
But it requires that we inventory where the classes are used, and for each use case, whenever it requires writing (else it should use interface instead).

### CODE REVIEW

**Comment by miurahr** [comment]
I agree that there are several code duplication in order to handle both TMXEntry and PrepareTMXEntry for same purpose, such as `searchEntreis` method. The PR here resolve it by common interface contract among two classes.

I found there is similar problem in two classes; TMXWrite class and TMXWrite2 class. 
Does the proposal also resolve the duplication on two classes?  TMXWrite handles PrepareTMXEntry and called from CLI methods. TMXWrite2 handles TMXEntry and called from GUI parts. Both class has same function to write project TMX file.

**Comment by miurahr** [comment]
You can take [this change for test cases from my PR](https://github.com/omegat-org/omegat/pull/184/commits/750b6a0594c0966e27a8a43b55d2fc4e2f211159) to avoid making TMXEntry/PrepareTMXEntry object variable, to resolve check error without any side effect.

**Comment by t-cordonnier** [comment]
> I found there is similar problem in two classes; TMXWrite class and TMXWrite2 class. 
> Does the proposal also resolve the duplication on two classes? 

Yes, removing all duplications is one of the goals of this work. But I maybe forgot to say that I did not consider it as finished, there is much more work to be done but I wanted to expose it to explain the principle
So yes, modifying the TMXWriter pair is in the next steps, I did not have the time to do it, I wanted to concentrate on other aspects before.

Today I published new changes. Now TMXEntry is not completely immutable: when a user wants to change the note of current translation, we modify it in TMXEntry rather than creating a new object each time. Second patch shows that I would like to do the same for setTranslation, but this one is only a prototype because I did not have the time to modify every method which calls it.
Once again, if there was a real reason why TMXEntry was immutable, I ask @amake  and @briacp  to explain why. Until that, I consider that modifying the object takes less time and memory than creating a new one on each modification.

**Comment by t-cordonnier** [comment]
> You can take [this change for test cases from my PR](https://github.com/omegat-org/omegat/pull/184/commits/750b6a0594c0966e27a8a43b55d2fc4e2f211159) to avoid making TMXEntry/PrepareTMXEntry object variable, to resolve check error without any side effect.

You are true I had completely forgotten the tests. Now I pushed a correction with --force.

**Comment by miurahr** [comment]
> > I found there is similar problem in two classes; TMXWrite class and TMXWrite2 class.
> > Does the proposal also resolve the duplication on two classes?
> 
> Yes, removing all duplications is one of the goals of this work. But I maybe forgot to say that I did not consider it as finished, there is much more work to be done but I wanted to expose it to explain the principle So yes, modifying the TMXWriter pair is in the next steps, I did not have the time to do it, I wanted to concentrate on other aspects before.

OK, I'd like to work for it in another PR without dependency for #184 and #185. 
When The PR #185 completed and provide backward compatibility,  my proposal can be still working.

**Comment by miurahr** [comment]
> > > I found there is similar problem in two classes; TMXWrite class and TMXWrite2 class.
> > > Does the proposal also resolve the duplication on two classes?
> > 
> > 
> > Yes, removing all duplications is one of the goals of this work. But I maybe forgot to say that I did not consider it as finished, there is much more work to be done but I wanted to expose it to explain the principle So yes, modifying the TMXWriter pair is in the next steps, I did not have the time to do it, I wanted to concentrate on other aspects before.
> 
> OK, I'd like to work for it in another PR without dependency for #184 and #185. When The PR #185 completed and provide backward compatibility, my proposal can be still working.

Done in #187

**Comment by miurahr** [comment]
> As @amake mentionned in the discussion list, PrepareTMXEntry actually looks like mutable variant of TMXEntry. That was initial meaning, but if you look where they are used, you will see that since ExternalTMX continued to use PrepareTMXEntry, it has been used as something similar to external TMX entry.
One other difference between these two classes is that PrepareTMXEntry can contain any kind of properties (as it can be read from any kind of TMX) while TMXEntry stores only properties a project_save.tmx should contain (actually no property but already stores externalLinked): this enforces the idea that TMXEntry means, in reality, internal TMX entry.
Then we can ask ourselves: why does ExternalTMX store mutable entries while internal (project) TMX contains immutable entries? Shouldn't the contrary be more logical?

I don't think it is good idea to keep using `PrepareTMXEntry` other than *Preparetion* class of `TMXEntry`. This is *the*  reason of my frustration. 

 A design which allow `PrepareTMX `as yet another container and allow getting values  from it bring unnecessary complexity in the project.
I'd like to propose an extension of TMXEntry which *can* have properties. Only an `ExternalTMFactory` want to bring it into `TMXEntry` object. Other parts can be as-is. This is what #188 doing. 

If we want to have  the class which is `TMXEntry` variant *with properties*, we can introduce `ExternalTMXEntry` for the purpose, that seems have a same ancestor with `TMXEntry`,  and make PrepareTMXEntry as for *preparation*.
There can be `ExternalTMXEntry extends TMXEntry` with properties, and `ExternalTMFactor` use it.

**Comment by t-cordonnier** [comment]
@miurahr Yes, the final idea (remember that I said my work is not finished) is something like ExternalTMXEntry with capacity to store variable list of properties while InternalTMXEntry would store only what a project_save.tmx is susceptible to contain. This is better than storing always maps because you don't need to store the same key thousands of times, and data can have a type (for example ExternalLinked)
In my logic PrepareTMXEntry will also disappear but I need some time to produce a patch, even as a prototype like the last one, to show you what I want to do as an alternative.
In any case the most important for me is the pair of interfaces: even if later the core team totally rejects the rest of the code, using these interfaces will ensure compatibility with later implementations.

**Comment by t-cordonnier** [comment]
As promised, I removed all work about (im)mutability but also introduced ExternalTMXEntry
Now both internal (project) and external TMX entries are immutable 
PrepareTMXEntry still exists, usage as ITMXEntry is permitted for compatibility with plugins only: in the code itself, it is used as a builder for Project/External TMX entries

**Comment by miurahr** [comment]
Looks good but CI test failed.

It looks good direction. I think there is no big problem for change core data in OmegaT OSS ecosystem, because there is no obvious effect/objection by plugin maintainers, so source compatibility may be enough.

**Comment by miurahr** [comment]
@amake could you rerun CI test? It seems  an issue in Azure-Pipeline service, not on us.

File: src/org/omegat/core/data/ExternalTMXEntry.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+ * unchangeable.
+ *
+ * Only RealProject can create and change TMXEntry objects.
+ *
+ * @author Alex Buloichik (alex73mail@gmail.com)
+ * @author Guido Leenders
+ * @author Aaron Madlon-Kay
+ */
+public class ExternalTMXEntry extends TMXEntry {
+    public List<TMXProp> otherProperties;
```
Comment (miurahr) [inline src/org/omegat/core/data/ExternalTMXEntry.java:49]: I think this can be and should be private.

File: src/org/omegat/core/data/ExternalTMXEntry.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(changeDate / 1000, creationDate / 1000, translation, note, changer, creator,
+                source);
+    }
+    
+    public List<TMXProp> getProperties() {
+        return otherProperties;
```
Comment (miurahr) [inline src/org/omegat/core/data/ExternalTMXEntry.java:92]: This should return a copy of properties. For example, 
```java
return Collections.unmodifiableList(otherProperties)
```

File: src/org/omegat/core/data/ImportFromAutoTMX.java [hunk 3]
```diff
@@ -93,48 +93,48 @@ void process(ExternalTMX tmx, boolean isEnforcedTMX) {
...
                     }
                 }
             }
         }
     }
 
-    private boolean isAltTranslation(PrepareTMXEntry entry) {
-        if (entry.otherProperties == null) {
+    private boolean isAltTranslation(ITMXEntry entry) {
+        if (entry.getProperties() == null) {
```
Comment (miurahr) [inline src/org/omegat/core/data/ImportFromAutoTMX.java:132]: we should introduce `boolean hasProperties()` method then we can encapsulate properties field in class.

File: src/org/omegat/core/data/ProjectTMXEntry.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
+/**************************************************************************
+ OmegaT - Computer Assisted Translation (CAT) tool
+          with fuzzy matching, translation memory, keyword search,
+          glossaries, and translation leveraging into updated projects.
+
+ Copyright (C) 2010 Alex Buloichik
+               2012 Guido Leenders, Thomas Cordonnier
+               2013 Aaron Madlon-Kay
+               2014 Alex Buloichik, Aaron Madlon-Kay
```
Comment (miurahr) [inline src/org/omegat/core/data/ProjectTMXEntry.java:9]: Your copyright here?

**Review by miurahr — CHANGES_REQUESTED** [review]
A total direction looks good but there are several point where you can update, and some suggestions.

File: src/org/omegat/core/data/ExternalTMXEntry.java (hunk 1 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/ExternalTMXEntry.java:49]: I did it like that only because all other fields remained public in the other classes. And setting other fields as private requires lot of changes in the code. But I agree that it can be discutable. Tell me what you think

File: src/org/omegat/core/data/ExternalTMXEntry.java (hunk 2 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/ExternalTMXEntry.java]: Agree, change pushed today

File: src/org/omegat/core/data/ImportFromAutoTMX.java (hunk 3 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/ImportFromAutoTMX.java]: Agree, change pushed today

File: src/org/omegat/core/data/ProjectTMXEntry.java (hunk 4 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/ProjectTMXEntry.java:9]: Agree, change pushed today

File: src/org/omegat/core/data/ExternalTMXEntry.java (hunk 1 above)
Comment (miurahr) [reply src/org/omegat/core/data/ExternalTMXEntry.java:51]: My idea is...
 As we discussed in developer mailing list, List<> is flexible but not optimal for memory consumption.
This is only chance to change access modifier for `otherProperties` because `ExternalTMXEntry` is new class. Other fields should public for backward compatibility.
 After we introduce `private otherProperties` and access methods, we can change it into other representation.
If we put it as `public`, we lost a chance to change it in future.

**Comment by miurahr** [comment]
There are over handful checkstyle warnings, actually 15, for new class `ExternalTMXEntry`.
You can take care for warnings when you introduced new code. I think we should care it at least when touched.
Same for ITMXEntry, `PrepareTMXEntry` getters, `ProjectTMXEntry` getters and `TMXEntry` getters.

OmegaT dev community dislike a style only patch, we should be careful and check checkstyle warnings when adding classes and methods.

File: src/org/omegat/core/data/TMXEntry.java [hunk 5]
```diff
@@ -28,51 +28,66 @@
 
 package org.omegat.core.data;
 
+import java.util.List;
```
Comment (miurahr) [inline src/org/omegat/core/data/TMXEntry.java:31]: This is unused import.

File: src/org/omegat/core/data/TMXEntry.java [hunk 6]
```diff
@@ -28,51 +28,66 @@
...
 import java.util.Objects;
 
+import org.omegat.util.TMXProp;
```
Comment (miurahr) [inline src/org/omegat/core/data/TMXEntry.java:34]: Unused import.

File: src/org/omegat/core/data/TMXEntry.java [hunk 7]
```diff
@@ -110,21 +125,12 @@ public boolean equals(Object obj) {
...
-            return false;
-        }
         if (!Objects.equals(source, other.source)) {
             return false;
         }
         return true;
     }
 
-    @Override
-    public int hashCode() {
```
Comment (miurahr) [inline src/org/omegat/core/data/TMXEntry.java:123]: When you define `equals` you should also define `hashCode`, this is one of checkstyle warning.

File: src/org/omegat/core/data/ITMXEntry.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+     * Gets text note (markup &lt;note&gt; in TMX format)
+     */
+    String getNote();
+    
+    /* --------------------- Properties ------------ */
+    
+    boolean hasProperties();
+    
+    String getPropValue(String propType);
+
```
Comment (miurahr) [inline src/org/omegat/core/data/ITMXEntry.java:70]: white space in empty line.

**Review by miurahr — CHANGES_REQUESTED** [review]
There is several obvious coding style issues.

**Comment by miurahr** [comment]
I've tried to merge https://github.com/omegat-org/omegat/pull/187  , I found it conflict with introduction of `ProjectTMXEntry`.

I need to add method `buildTMXFile` that take `Map<String, TMXEntry>` as source data, that need to change `Map<String, ProjectTMXEntry>` but package-private `PrepareTMXEntry#toProjectTMXEntry` can not be looked from Main class.

```java
    public static void buildTMXFile(final String filename, final boolean forceValidTMX,
                                    final boolean levelTwo, final ProjectProperties config,
                                    final Map<String, ProjectTMXEntry> data) throws Exception {
        TMXWriter2 wr = new TMXWriter2(new File(filename), config.getSourceLanguage(), config.getTargetLanguage(),
                config.isSentenceSegmentingEnabled(), levelTwo, forceValidTMX);
        List<String> p = new ArrayList<>();
        try {
            for (Map.Entry<String, ProjectTMXEntry> en : data.entrySet()) {
                TMXEntry transEntry = en.getValue();
                wr.writeEntry(en.getKey(), transEntry.getTranslationText(), transEntry.getNote(),
                        transEntry.getCreator(), transEntry.getCreationDate(), transEntry.getChanger(),
                        transEntry.getChangeDate(), p);
            }
        } finally {
            wr.close();
        }
    }
```

File: src/org/omegat/core/data/PrepareTMXEntry.java [hunk 9]
```diff
@@ -107,5 +147,12 @@ public String toString() {
                 .append(", otherProperties=").append(otherProperties).append("]");
         return builder.toString();
     }
+    
+    ExternalTMXEntry toExternalTMXEntry() {
+        return new ExternalTMXEntry(this);
+    }
 
+    ProjectTMXEntry toProjectTMXEntry(boolean defaultTranslation, ProjectTMXEntry.ExternalLinked linked) {
```
Comment (miurahr) [inline src/org/omegat/core/data/PrepareTMXEntry.java:155]: Could this change to `public` because  [TMXWriter2 improvement conflicts with this](https://github.com/omegat-org/omegat/pull/185#issuecomment-1017136973)

Comment (miurahr) [reply src/org/omegat/core/data/PrepareTMXEntry.java:155]: There is another idea to introduce `static ProjectTMX#createPseudoTranslateTMX` in `org.omegat.core.data` package instead of change to `public`.

**Comment by t-cordonnier** [comment]
> public static void buildTMXFile(final String filename, final boolean forceValidTMX,
>                                    final boolean levelTwo, final ProjectProperties config,
>                                    final Map<String, ProjectTMXEntry> data) 
For me this method should take Map<String, ? extends ITMXEntry> data instead: now we should try to have methods as generic as possible
> List<String> p = new ArrayList<>();
> [...]
> wr.writeEntry(en.getKey(), transEntry.getTranslationText(), transEntry.getNote(),
> [...]
 >                                               transEntry.getChangeDate(), p);
Strange implementation here. For genericity, this method should take the properties from the entry itself (now ITMXEntry has properties method, which TMXEntry did not)

You may say that there are situations where we want to write other properties than the ones of the entry itself: for example project_save.tmx contains key properties which are not registered in ProjectTMXEntry. I will study this case but I suppose that we can use the already existing method in this case:
    public void writeEntry(String source, String translation, String note, String creator, long creationDate,
            String changer, long changeDate, List<String> propValues)

File: src/org/omegat/core/data/PrepareTMXEntry.java (hunk 9 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/PrepareTMXEntry.java:155]: This method is package-protected because previous constructor TMXEntry(PrepareTMXEntry) was like that and the comments in the top of this class said that it was intentional.
Now look to my answer to your previous comment and tell me if it is still required

**Comment by miurahr** [comment]
Thank you for the suggestion. I'd like to move the TMX writing logic in Main class,  and remove dependency to  `ProjectTMXEntry` class.

**Comment by t-cordonnier** [comment]
Yes I am working on it, I will add a new patch when I have time

**Comment by miurahr** [comment]
Similar issue on `runConsoleAlign`:
It can be 

```        TMXWriter2 wr = new TMXWriter2(new File(tmxFile), config.getSourceLanguage(), config.getTargetLanguage(),
                config.isSentenceSegmentingEnabled(), false, false);
        Map<String, ProjectTMXEntry> data = p.align(p.getProjectProperties(), new File(dir));
        try {
            for (Map.Entry<String, ProjectTMXEntry> en : data.entrySet()) {
                wr.writeEntry(en.getKey(), en.getValue().getTranslationText(), en.getValue(), null);
            }
        } finally {
            wr.close();
        }
```

It looks clean now.

**Comment by miurahr** [comment]
I've updated #187 .
Here is an updated merge https://github.com/miurahr/omegat/blob/topic/t-cordonnier/refactor-tmxEntry/src/org/omegat/Main.java#L447-L508
please check and compare with yours.

File: src/org/omegat/core/data/RealProject.java [hunk 10]
```diff
@@ -1814,7 +1814,7 @@ protected String getSegmentTranslation(String id, int segmentIndex, String segme
             this.config = props;
         }
 
-        Map<String, TMXEntry> data = new HashMap<>();
+        Map<String, ProjectTMXEntry> data = new HashMap<>();
```
Comment (miurahr) [inline src/org/omegat/core/data/RealProject.java:1817]: Let's change this to TreeMap, based on discussion https://github.com/omegat-org/omegat/commit/e740a9753201228a08048a3296377583ac2312bd

**Comment by miurahr** [comment]
Test failed because there is a missing update for integration test code.
```
test-integration/src/org/omegat/core/data/TestTeamIntegrationChild.java:668: error: incompatible types: TMXEntry cannot be converted to ProjectTMXEntry
            mergedTMX.setTranslation(ste, en, true);
```

**Comment by t-cordonnier** [comment]
> Test failed because there is a missing update for integration test code.

Correct, should be ok now

File: src/org/omegat/core/data/TMXEntry.java (hunk 7 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/TMXEntry.java:123]: This warning is a little bit stupid: the method has been moved to concrete classes (Project,External)TMXEntry because the implementation differs, due to different internal structure.
Should we define an abstract method only to make the style checker happy?

Comment (miurahr) [reply src/org/omegat/core/data/TMXEntry.java:123]: You can add `@SuppressWarnings` annotation here

File: src/org/omegat/gui/glossary/GlossaryEntry.java [hunk 11]
```diff
@@ -58,8 +59,18 @@ public GlossaryEntry(String src, String loc, String com, boolean fromPriorityGlo
 
     public String getSrcText() {
         return mSource;
-    }
+    }
```
Comment (miurahr) [inline src/org/omegat/gui/glossary/GlossaryEntry.java:62]: Codestyle: spaces on end of line

File: src/org/omegat/core/data/PrepareTMXEntry.java [hunk 12]
```diff
@@ -33,18 +33,22 @@
...
 /**
- * Class for prepare TMXEntry content before save unchangeable copy in the ProjectTMX. We can't use just
+ * Class for prepare TMXEntry content before save unchangeable copy anywhere. We can't use just
  * parameters in the setTranslation() method since count of parameters is too much. Structure of this class is
  * almost the save like TMXEntry.
  *
  * Instead, we will set all parameters into this class, then ProjectTMX will convert in into TMXEntry than
  * save internally.
  *
+ * In the past, entries read from tm/ folder were kept as instances of PrepareTMXEntry.
```
Comment (miurahr) [inline src/org/omegat/core/data/PrepareTMXEntry.java:43]: codingstyle:  Line has trailing spaces.

File: src/org/omegat/core/data/ProjectTMXEntry.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    ProjectTMXEntry(ITMXEntry from, boolean defaultTranslation, ExternalLinked linked) {
+        super(from);
+
+        this.defaultTranslation = defaultTranslation;
+        this.linked = linked;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (! super.equals(obj)) {
```
Comment (miurahr) [inline src/org/omegat/core/data/ProjectTMXEntry.java:66]: Many checkstyle warnings to be fixed to fix based on coding guide on OmegaT.
```
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:64:5: Class 'ProjectTMXEntry' looks like designed for extension (can be subclassed), but the method 'equals' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'ProjectTMXEntry' final or making the method 'equals' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:66:14: '!' is followed by whitespace. [NoWhitespaceAfter]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:80:5: Class 'ProjectTMXEntry' looks like designed for extension (can be subclassed), but the method 'hashCode' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'ProjectTMXEntry' final or making the method 'hashCode' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:82:42: '1000' is a magic number. [MagicNumber]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:82:63: '1000' is a magic number. [MagicNumber]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:91:14: '!' is followed by whitespace. [NoWhitespaceAfter]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:104: Line has trailing spaces. [RegexpSingleline]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:105:5: Class 'ProjectTMXEntry' looks like designed for extension (can be subclassed), but the method 'hasProperties' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'ProjectTMXEntry' final or making the method 'hasProperties' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:108: Line has trailing spaces. [RegexpSingleline]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:109:5: Class 'ProjectTMXEntry' looks like designed for extension (can be subclassed), but the method 'getPropValue' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'ProjectTMXEntry' final or making the method 'getPropValue' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:112: Line has trailing spaces. [RegexpSingleline]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:113:5: Class 'ProjectTMXEntry' looks like designed for extension (can be subclassed), but the method 'hasPropValue' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'ProjectTMXEntry' final or making the method 'hasPropValue' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:116: Line has trailing spaces. [RegexpSingleline]
[ant:checkstyle] [WARN] /home/vsts/work/1/s/src/org/omegat/core/data/ProjectTMXEntry.java:117:5: Class 'ProjectTMXEntry' looks like designed for extension (can be subclassed), but the method 'getProperties' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'ProjectTMXEntry' final or making the method 'getProperties' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
```

**Review by miurahr — CHANGES_REQUESTED** [review]
There are new raising checkstyle warnings.
Please take care  especially for added new classes.

File: src/org/omegat/core/data/ProjectTMXEntry.java (hunk 13 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/ProjectTMXEntry.java:66]: I converted the PR to draft so that we stop concentrate on checkstyle; I have still some points to test before considering everything is correct, and come back to the style before moving again to a real pull request

File: src/org/omegat/core/data/TMXEntry.java (hunk 5 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/TMXEntry.java:31]: Delayed, was true in a patch which will be moved to another pull request

File: src/org/omegat/core/data/TMXEntry.java (hunk 7 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/TMXEntry.java]: Patched moved to other PR, to be considered later

File: src/org/omegat/gui/glossary/GlossaryEntry.java (hunk 11 above)
Comment (t-cordonnier) [reply src/org/omegat/gui/glossary/GlossaryEntry.java]: Solved

File: src/org/omegat/core/data/PrepareTMXEntry.java (hunk 12 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/PrepareTMXEntry.java]: Patch moved to other PR

File: src/org/omegat/core/data/ProjectTMXEntry.java (hunk 13 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/ProjectTMXEntry.java]: Definition of ProjectTMXEntry moved to another PR

**Comment by t-cordonnier** [comment]
Regarding the problem of backward or upward compatibility with plugins, I had sent the results of my tests here:
https://sourceforge.net/p/omegat/mailman/message/37593315/

I did also test against my translation memories API and come to exactly same conclusion: the 3 first patches have zero impact on plugins, 3 last patches can have an impact.

For all that reasons, I remove the draft status for the 3 first patches, and will open a new pull request with other patches, which can be treated later. Only problem is that GitHub provides no way to make clear that the second PR depends on this one.

File: src/org/omegat/core/data/ImportFromAutoTMX.java [hunk 14]
```diff
@@ -169,20 +169,14 @@ private boolean altTranslationMatches(PrepareTMXEntry entry, EntryKey key) {
...
-        if (StringUtil.isEmpty(trans.note)) {
-            trans.note = null;
-        }
-
-        trans.source = entry.getSrcText();
-
         TMXEntry newTrEntry;
 
-        if (trans.translation == null && trans.note == null) {
+        if (trans.getTranslationText() == null && (trans.getNote() == null || trans.getNote().length() == 0)) {
```
Comment (amake) [inline src/org/omegat/core/data/ImportFromAutoTMX.java:179]: I would be OK with a `hasNote()` method that computes `note != null && !note.isEmpty()`

**Review by amake — APPROVED** [review]
LGTM! I have given @t-cordonnier and @miurahr write access to the SourceForge git repo, so please push when you are ready.

File: src/org/omegat/core/data/ImportFromAutoTMX.java (hunk 14 above)
Comment (t-cordonnier) [reply src/org/omegat/core/data/ImportFromAutoTMX.java]: OK, indeed the method was in TMXEntry, so I have moved it to the interface so that we are sure to have a consistent behaviour everywhere we would need it.
(in particular this line was not correct since it seems that OmegaT considers that empty note is different from no note at all, so now it is corrected)

**Review by miurahr — APPROVED** [review]

**Comment by miurahr** [comment]
LGTM. I think it is ready to merge.

**Comment by t-cordonnier** [comment]
@amake, @miruahr
Thanks for your review and for writing permissions
One last point before validating this PR

As you mentioned previously the checkstyle results, I see that the new methods often add the following warning:
"Class 'PrepareTMXEntry' looks like designed for extension (can be subclassed), but the method 'getPropValue' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'PrepareTMXEntry' final or making the method 'getPropValue' static/final/abstract/empty, or adding allowed annotation for the method."
Setting the methods as final makes this warning disappear.
However, while it sounds clear for me that method like getCreator() or getNote() can be final, some other methods like getPropValue(String)  may be overridden, in particular in case of TMXEntry (look in PR #202, even if I know that this is a draft with errors, for example)
What makes also the warning to disappear is to add any kind of javadoc, such as /** do what you want **/

I also notice that this warning is present in lot of classes, and if we take care of it, it will be the first time.

So, the question is, should I
- ignore and validate as is
- add final attribute where I think it is correct (in accessors, but not in methods which may be modified)
- add a javadoc in methods which cannot be final, only to make the checkstyle happy, even if I don't know what to say...

Tell me what you think and we can validate the PR then.

**Comment by amake** [comment]
My stance on CheckStyle is that formatting-level things like spaces and line length should be fixed unless there is a good reason not to.

On the other hand, things that affect semantics can generally be ignored. I have considered disabling the "designed for extension" rule entirely, but I left it on because it's sometimes good food for thought.

In this case I think

> - ignore and validate as is

is fine.

**Comment by t-cordonnier** [comment]
Since both Aaron and Hiroshi confirmed, I did the merge.
Thanks to Aaron who gave me rights to do so

**Comment by miurahr** [comment]
@t-cordonnier you merge here to github's mirror,  not source-forge's our origin source of truth.

@amake can I push mirror onto origin w/o rebase ?

**Comment by miurahr** [comment]
I've just pushed PR #208 to express merge strategy and procedure in docs_devel.

**Comment by t-cordonnier** [comment]
> @t-cordonnier you merge here to github's mirror, not source-forge's our origin source of truth.

Thanks for the info. Indeed I only clicked on the button in GitHub's web interface, was there another procedure?

**Comment by amake** [comment]
> Thanks for the info. Indeed I only clicked on the button in GitHub's web interface, was there another procedure?

It must not have been clear, but I thought I had mentioned that the source of truth is the SourceForge git repo. The GitHub mirror is synced on-demand, so any changes made only here will be overwritten.

I've squashed your changes to SourceForge, and they have been synced here now: https://github.com/omegat-org/omegat/commit/569ebf2f8cc15f864354afdc456fadf24555097c
